python/cuml/dask/common/dask_arr_utils.py

```diff
@@ -0,0 +1,180 @@
+        worker_map[key_to_part_dict[key]] = worker
+
+    worker_to_parts = []
+    for part in parts:
+        worker = worker_map[part]
+        worker_to_parts.append((worker, part))
+
+    yield wait(worker_to_parts)
+    raise gen.Return(worker_to_parts)
+
+
+def _x_p(x):
+    return x
+
+
+def _conv_np_to_df(x):
+    cupy_ary = rmm_cupy_ary(cp.asarray,
+                            x,
+                            dtype=x.dtype)
+    return cudf.DataFrame.from_gpu_matrix(cupy_ary)
+
+
+def _conv_df_to_sp(x):
+    cupy_ary = rmm_cupy_ary(cp.asarray,
+                            x.as_gpu_matrix(),
+                            dtype=x.dtypes[0])
+
+    return cp.sparse.csr_matrix(cupy_ary)
+
+
+def to_sp_dask_array(cudf_or_array, client=None):
+    """
+    Converts an array or cuDF to a sparse Dask array backed by sparse CuPy.
+    CSR matrices. Unfortunately, due to current limitations in Dask, there is
+    no direct path to convert a cupy.sparse.spmatrix into a CuPy backed
+    dask.Array without copying to host.
+
+
+    NOTE: Until https://github.com/cupy/cupy/issues/2655 and
+    https://github.com/dask/dask/issues/5604 are implemented, compute()
+    will not be able to be called on a Dask.array that is backed with
+    sparse CuPy arrays because they lack the necessary functionality
+    to be stacked into a single array. The array returned from this
+    utility will, however, still be able to be passed into functions
+    that can make use of sparse CuPy-backed Dask.Array (eg. Distributed
+    Naive Bayes).
+
+    Relevant cuML issue: https://github.com/rapidsai/cuml/issues/1387
+
+    Parameters
+    ----------
+    cudf_or_array : cuDF, array-like sparse / dense array, or
+                    Dask DataFrame/Array
+    client : dask.distributed.Client (optional) Dask client
+
+    dtype : output dtype
+
+    Returns
+    -------
+    dask_array : dask.Array backed by cupy.sparse.csr_matrix
+    """
+    client = default_client() if client is None else client
+
+    shape = cudf_or_array.shape
+    if isinstance(cudf_or_array, dask.dataframe.DataFrame) or \
+       isinstance(cudf_or_array, cudf.DataFrame):
+        dtype = cudf_or_array.dtypes[0]
+    else:
+        dtype = cudf_or_array.dtype
+
+    meta = cp.sparse.csr_matrix(rmm_cupy_ary(cp.zeros, 1))
+
+    if isinstance(cudf_or_array, dask.array.Array):
+        # At the time of developing this, using map_blocks will not work
+        # to convert a Dask.Array to CuPy sparse arrays underneath.
+        parts = client.sync(extract_arr_partitions, cudf_or_array)
+        cudf_or_array = [client.submit(_conv_np_to_df, part, workers=[w])
+                         for w, part in parts]
+
+        cudf_or_array = to_dask_cudf(cudf_or_array)
+
+    if isinstance(cudf_or_array, dask.dataframe.DataFrame):
+        """
+        Dask.Dataframe needs special attention since it has multiple dtypes.
+        Just use the first (and assume all the rest are the same)
```

Comment:
Could we have a gh issue to look into having a check that they are indeed the same?

Reply:
We could do a unique here just to verify. That Would work